src/Aquifer.Common/Services/Caching/CachingApiKeyService.cs

```diff
@@ -1,51 +1,73 @@
     private static readonly TimeSpan s_cacheLifetime = TimeSpan.FromMinutes(10);
 
-    public async Task<CachedApiKey?> GetCachedApiKeyAsync(string apiKey, CancellationToken ct)
+    public ApiKey CurrentApiKey
     {
-        return (await GetApiKeysFromCacheAsync(ct)).SingleOrDefault(x => x.ApiKey == apiKey);
+        get => _httpContextAccessor.HttpContext?.Items[Constants.HttpContextItemCachedApiKey] as ApiKey
+            ?? throw new InvalidOperationException("API Key should be saved on HTTP Context!");
+        set => _httpContextAccessor.HttpContext?.Items.Add(Constants.HttpContextItemCachedApiKey, value);
     }
 
-    public async Task<CachedApiKey?> GetCachedApiKeyAsync(string apiKey, ApiKeyScope scope, CancellationToken ct)
+    public async Task<ApiKey?> GetApiKeyAsync(string apiKey, CancellationToken ct)
     {
-        return (await GetApiKeysFromCacheAsync(ct)).SingleOrDefault(
-            x => (x.Scope == scope || x.Scope == ApiKeyScope.All) && x.ApiKey == apiKey);
+        return (await GetApiKeysFromCacheAsync(ct)).SingleOrDefault(x => x.Key == apiKey);
+    }
+
+    public async Task<ApiKey?> GetApiKeyAsync(string apiKey, ApiKeyScope scope, CancellationToken ct)
+    {
+        return (await GetApiKeysFromCacheAsync(ct)).SingleOrDefault(x => x.Key == apiKey && x.HasScope(scope));
     }
 
     public async Task<bool> IsValidApiKeyAsync(string apiKey, ApiKeyScope scope, CancellationToken ct)
     {
-        var scopedApiKey = await GetCachedApiKeyAsync(apiKey, scope, ct);
+        var scopedApiKey = await GetApiKeyAsync(apiKey, scope, ct);
         return scopedApiKey is not null;
     }
 
-    private async Task<List<CachedApiKey>> GetApiKeysFromCacheAsync(CancellationToken ct)
+    private async Task<List<ApiKey>> GetApiKeysFromCacheAsync(CancellationToken ct)
     {
         return await _memoryCache.GetOrCreateAsync(
                 ApiKeysCacheKey,
                 async cacheEntry =>
                 {
                     cacheEntry.AbsoluteExpirationRelativeToNow = s_cacheLifetime;
 
-                    return await _dbContext.ApiKeys.Select(x => new CachedApiKey(x.Id, x.ApiKey, x.Scope, x.Organization)).ToListAsync(ct);
+                    return await _dbContext.ApiKeys.Select(x => new ApiKey(x.Id, x.ApiKey, x.Scope, x.Organization)).ToListAsync(ct);
                 }) ??
             throw new InvalidOperationException($"{ApiKeysCacheKey} unexpectedly had a null value cached.");
     }
 }
 
-public record CachedApiKey(int Id, string ApiKey, ApiKeyScope Scope, string? Organization);
\ No newline at end of file
+public sealed class ApiKey(int _id, string _key, ApiKeyScope _scope, string? _organization)
+{
+    public int Id { get; } = _id;
+    public string Key { get; } = _key;
+    public ApiKeyScope Scope { get; } = _scope;
+    public string? Organization { get; } = _organization;
+
+    public bool HasScope(ApiKeyScope scope)
+    {
+        return scope is ApiKeyScope.Admin or ApiKeyScope.None
```

Comment:
I wouldn't expect `All` or `Admin` to match a scope of `None` so I put an exception for it in here.